src/Illuminate/Collections/helpers.php

```diff
@@ -163,13 +163,13 @@ function data_forget(&$target, $key)
 
         if (($segment = array_shift($segments)) === '*' && Arr::accessible($target)) {
             if ($segments) {
-                foreach ($target as &$inner) {
-                    data_forget($inner, $segments);
+                foreach ($target as $segment => &$inner) {
+                    $target[$segment] = data_forget($inner, $segments);
                 }
             }
         } elseif (Arr::accessible($target)) {
-            if ($segments && Arr::exists($target, $segment)) {
-                data_forget($target[$segment], $segments);
+            if ($segments && $inner = $target[$segment]) {
```

Comment:
Behavior change here. Arr::exist would return true for `null`. But, your change would seem to consider `null` as falsy.

Reply:
Yeah, fair point. I had done some testing with it and figured that it would end up as approximately the same result as a falsy value would fail the conditional and flow through to `Arr::forget($target, $segment);` (thus being unset), but I can see that that's not entirely valid.

```php
$data = ['one' => ['two' => null]];
data_forget($data, 'one.two.three');
```
would result in `['one' => []]` instead of leaving `two` alone (because it's only supposed to remove the `three` option within `two`).

I'll push up a fix for this, and I'll expand the tests to cover it.

Incidentally (and I'm not going to resolve it in this PR, but probably worth considering) - what would you expect the behaviour of `data_forget($something, '*')` to be? It's not something that I _expect_ to be done in userland ever, but currently it will not do anything at all. I would expect that it would just empty the item that you passed, but currently it won't. This also affects recursive unsets `data_forget($something, 'foo.bar.*')`.

It's a pretty easy fix to make it actually clear out the object, but that's certainly a different PR.